Finalize gateway API

VoidingWarranties opened the pull request:
add & remove isn't as clear as connect & disconnect, especially since the peer remains in the node list on disconnect. add & remove could be used in the future to add and remove addresses from the node list.

lukechampine commented on siac/gatewaycmd.go:
```diff
@@ -14,22 +14,36 @@ var (
+	gatewayDeprecatedRemoveCmd = &cobra.Command{
+		Use:        "remove [address]",
+		Deprecated: "use `siac gateway disconnect` instead.",
+		Run:        wrap(gatewaydisconnectcmd),
```
I think you can use an alias here, like we do for `renter ls`

VoidingWarranties replied:
If we want to remove the `remove` command eventually we should deprecate it instead of aliasing it so that the user is warned that its deprecated when they try to use it.

DavidVorick commented on api/api.go:
```diff
@@ -115,8 +115,11 @@ func (srv *Server) initAPI(password string) {
+		router.POST("/gateway/disconnect/:netaddress", requirePassword(srv.gatewayDisconnectHandler, password))
+		// COMPATv0.6.0
+		router.POST("/gateway/add/:netaddress", requirePassword(srv.gatewayConnectHandler, password))
+		router.POST("/gateway/remove/:netaddress", requirePassword(srv.gatewayDisconnectHandler, password))
```
When we switch to 1.0 we should remove all COMPAT calls in the api

DavidVorick replied:
This PR is intended to be a part of preparing for 1.0... remove these calls :P

DavidVorick commented:
As a part of this pull request, I would like to see a file created in `doc/api` called `Gateway.md` which details all of the gateway calls and what they do. They should document all of the input fields, all of the output fields, and as you create this file you should spellcheck everything.

Don't forget to spellcheck objects that get used. For example, in the host there's the `HostInternalSettings` object, which has JSON that is unnamed in the api package. To spellcheck the API's `HostInternalSettings` related calls, you'd have to look at the code in the `modules` package where `HostInternalSettings` is defined.

VoidingWarranties commented:
How do you want `doc/api/Gateway.md` to be different from the gateway section in `doc/API.md`?

DavidVorick commented:
Sorry for not answering the question earlier. In an attempt to finalize the host api, I've got an example, checkout the documentation changes in #1234 

I'm not particularly attached to any method, except that I insist that the whole project be consistent.

There are a few major differences to the way that the host api doc has been structured:

+ Provide a list of host routes in API.md, along with a list of fields and parameters, but not the full description or explanation.
+ Link to the full description
+ Put the full description in doc/api/Host.md instead of in doc/API.md
+ Explain what a field does inline instead of explaining it at the bottom. The documentation for a field is now closer to the field, which hopefully improves readability.

A table of contents sounds reasonable.

The way I've set up the host means that there's one reference sheet with all of the routes and their fields, but it's very dense (I think this is a good thing). The full descriptions, if you need them, are available in separate files.

Benefits: breaks out a full host.md description,

VoidingWarranties commented:
Updated the gateway section of `doc/API.md` and added `doc/api/Gateway.md`. I thought a lot about the layout of these, so please let me know if you have any thoughts.

VoidingWarranties commented:
Notably the structs have changed from Golang syntax to JSON (really javascript) syntax. Clients of the API are unlikely to be using Golang, and the API docs should be language agnostic.

Also thanks to @johnathanhowell for providing feedback offline.

DavidVorick commented on doc/API.md:
```diff
@@ -27,6 +27,32 @@ Example GET curl call:  `curl -A "Sia-Agent" /wallet/transactions?startheight=1&
+#### Error
+
+The standard error response indicating the request failed for any reason, is a
+4xx HTTP status code with the error message as plain text.
```
I believe that the error message should be returned in json, that way additional objects can be added for compat support or other weirdness.

This isn't the link I was thinking of, but it does talk about error design:  https://apigee.com/about/blog/technology/restful-api-design-what-about-errors

VoidingWarranties replied:
Here's a long discussion about JSON error design. https://github.com/json-api/json-api/issues/7

DavidVorick commented on doc/API.md:
```diff
@@ -27,6 +27,32 @@ Example GET curl call:  `curl -A "Sia-Agent" /wallet/transactions?startheight=1&
+- [Renter](#renter)
+- [Transaction Pool](#transaction-pool)
+- [Wallet](#wallet)
+
```
Table of contents looks good

DavidVorick commented on doc/API.md:
```diff
@@ -261,60 +287,66 @@ provided hash.
+| ----- | --------- |
+| [/gateway](#gateway-get) | GET |
+| [/gateway/connect/{netaddress}](#gatewayconnectnetaddress-post) | POST |
+| [/gateway/disconnect/{netaddress}](#gatewaydisconnectnetaddress-post) | POST |
```
Ideally this would be cleaner in markdown, can you align the bars? I believe it will still format correctly.

DavidVorick commented on doc/api/Gateway.md:
````diff
@@ -0,0 +1,163 @@
+}
+```
+
+###### [Example](#gateway-info-example)
````
The doc is really busy. I like most of the information, I like the examples and stuff, but there are a lot of elements and it's confusing to navigate.

One of the things that might help is to move the link to the example up to the line with the route definition. e.g.

`#### /gateway [GET] [(Example)](#gateway-info-example)`

DavidVorick replied:
![2016-06-15-200926_1789x1320_scrot](https://cloud.githubusercontent.com/assets/2316841/16101520/931862ac-3334-11e6-82f2-ba0c0561c9f4.png)

DavidVorick replied:
line 1: bold
line 2:  normal, with `code` at the front
line 3: faded bold
line 4-12: block
line 13: link
line 14: bold
line 15-16: normal, leading `with code`
line 17: faded bold
line 18 `code`
line 19: faded bold
line 20: normal
line 21: link

In 21 lines you've got 12 format changes. That's really difficult for me to read. 

From what I can tell, each route currently has a maximum of 7 different style changes:

route
route description
parameters title
parameters javascript
response title
response javascript / text
example

Ideally we could get this down to 3 or 4. I'm not sure we can get that low, but maybe you've got ideas. Perhaps:

route + example link
description
parameters javascript
response javascript

DavidVorick commented on doc/api/Gateway.md:
```diff
@@ -0,0 +1,163 @@
+
+* [gateway info](#gateway-info-example)
+* [connecting to a peer](#connect-example)
+* [disconnecting from a peer](#disconnect-example)
```
If you are planning on having an example for every route, you probably should not link to them all. For the gateway it's not so bad because there are only three, but it does add to the cluttered feeling.

Or maybe you can add an example field to the index table

```
| Route | HTTP verb | Example |
| ----- | --------- | ------- |
| [/gateway](#gateway-get) | GET | [gateway info](#gateway-info-example) |
| [/gateway/connect/{netaddress}](#gatewayconnectnetaddress-post) | POST | [connecting to a peer](#connect-example) |
| [/gateway/disconnect/{netaddress}](#gatewaydisconnectnetaddress-post) | POST | [disconnecting from a peer](#disconnect-example)| 
```

VoidingWarranties replied:
I think there's an advantage of having the example link next to each route description (or in the title), as a developer is more likely to want to look at an example when they are looking through the request or response specs.

DavidVorick commented on siac/main.go:
```diff
@@ -238,10 +238,14 @@ func main() {
 
+	// COMPATv0.6.0
+	gatewayCmd.AddCommand(gatewayDeprecatedAddCmd)
+	gatewayCmd.AddCommand(gatewayDeprecatedRemoveCmd)
```
remove compatibility code. 1.0 is brutal :)

DavidVorick replied:
It's fine to do it this way because people are expecting a moderate to substantial change with this release.

DavidVorick commented:
Overall the documentation looks good, there's a lot of good information and I like that we now have examples, but the way that the information is presented is still pretty cluttered and confusing.

Given that we're working in markdown, there's only so much we can do to address that. It's probably going to be cluttered no matter what. But hopefully we can get a step or two better than where it's at.

Either way, definitely a big improvement over where it was before.

VoidingWarranties commented:
Updated with better styling, removed COMPAT code, and JSON errors.
